File: 1126/main.py
dp = dict()

n = int(input())

heights = list(map(int, input().split()))

sh = [0]

for i in heights:
    sh.append(sh[-1] + i)
sh.pop(0)


maxdiff = sum(heights)//2

def findh(j,hd):
    if (j,hd) in dp.keys():
        return dp[(j,hd)]
    if hd > sh[j]:
        dp[(j,hd)] = -987654321
        return -987654321
    if hd > maxdiff:
        dp[(j,hd)] = -987654321
        return -987654321
    if j==0:
        if hd == heights[0]:
            return heights[0]
        else:
            dp[(j,hd)] = -987654321
            return -987654321
    curblock = heights[j]
    a = findh(j-1,hd)
    b = findh(j-1,hd+curblock)
    if hd >= curblock:
        c = findh(j-1,hd-curblock)+curblock
    else:
        c = findh(j-1,curblock-hd)+hd

    dp[(j,hd)] = max(a,b,c)
    return dp[(j,hd)]

ans = findh(n-1,0)
if ans >= 0:
    print(ans)
else:
    print(-1)

File: 1126/main.c
#include <stdio.h>

int dp[51][500001];

int sh[50];
int heights[50];

int findh(int j, int hd){
    if (j==-1){
        if(hd==0) return 0;
        return -987654321;
    }
    if (dp[j][hd] != -1){
        return dp[j][hd];
    }
    if (hd > sh[j]){
        return -987654321;
    }
    int a,b,c;
    a = findh(j-1, hd);
    b = findh(j-1,hd+heights[j]);
    if (hd > heights[j]){
        c = findh(j-1,hd-heights[j])+heights[j];
    }else{
        c = findh(j-1,heights[j]-hd)+hd;
    }
    int maxi = (a>b)?a:b;
    maxi = (maxi>c)?maxi:c;
    dp[j][hd] = maxi;
    return maxi;
}

int main(){
    int n, shbef;
    scanf("%d", &n);
    shbef = 0;
    for (int i=0;i<n;i++){
        scanf("%d",&heights[i]);
        sh[i] = shbef+heights[i];
        shbef = sh[i];
    }
    for(int i=0;i<51;i++){
        for(int j=0;j<500001;j++){
            dp[i][j] = -1;
        }
    }
    int ans = findh(n-1,0);
    if (ans > 0){
        printf("%d\n", ans);

    }else{
        printf("-1\n");
    }
    return 0;
}
